mongo/description/server_selector.go

```diff
@@ -296,13 +296,16 @@ func selectByTagSet(candidates []Server, tagSets []tag.Set) []Server {
 }
 
 func selectByKind(candidates []Server, kind ServerKind) []Server {
-	var result []Server
-	for _, s := range candidates {
+	index := make([]int, 0, len(candidates))
+	for i, s := range candidates {
 		if s.Kind == kind {
-			result = append(result, s)
+			index = append(index, i)
 		}
 	}
-
+	result := make([]Server, len(index))
+	for i, idx := range index {
```

Comment:
Do we need to do this second loop? Setting the slice capacity is pretty clever, can we just use it like this?

```go
results := make([]int, 0, len(candidates))
for _, s := range candidates {
    if s.Kind == kind {
        result s = append(results, s)
    }
}
return results
```

Or are we concerned about future relocations to the resulting slice?

Reply:
The major concern is how many candidates need to be appended. If only a very small part of the candidates will be appended, pre-allocating a slice of `Server` with a large capacity will be costly. (Maybe, I need to add this situation into our benchmark as well.) Pre-allocating a slice of `int` is less expensive and a better balance in my idea.

Reply:
I think it's fine either way, but do you expect enough servers needing to be filtered to justify this index optimization?

Reply:
In the use case described in the ticket, the reporter has hundreds of instances and all are `mongos` type. Either way works in that case. While we also filter `RSPrimary` and `RSSecondary`, I'm trying to optimize an ultimate scenario without de-optimize too much on the other end.